To find fonts that can represent a wide range of Unicode characters, we need to analyze the Unicode coverage of multiple fonts and compare them. 

Initial plan:
1. Create a list of fonts we want to evaluate.<newline>
2. Retrieve the supported Unicode character range for each font in the list.<newline>
3. Compile the results and compare the Unicode coverage among the fonts. Look for fonts that have a broader range of supported Unicode characters.<newline>
4. Consider factors such as the scripts or languages we’re interested in, the specific Unicode blocks we want to cover, and the visual design or stylistic preferences we have for the font.<newline>


In [1]:
#pip install fonttools

# Make sure you install this first!

In [2]:
from fontTools.ttLib import TTFont
import glob
import os

The code below uses the getBestCmap() method to obtain the mapping of Unicode characters supported by the font, and then it prints the Unicode characters in hexadecimal representation. We can also extract additional information about the font, such as the font family name and font style. The first font used is NotoSans, a popular choice as it aims to provide comprehensive coverage of all Unicode characters.

# 1. NotoSans

In [3]:
# Specify the path to the font file
font_path = "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/NotoSans-Regular.ttf"

# Load the font using TTFont
font = TTFont(font_path)

# Retrieve the supported Unicode character range
supported_range = font.getBestCmap().keys()

# # Print the supported Unicode characters
# for char in supported_range:
#     print(hex(char))

In [4]:
# Specify the path to the font file
font_path = "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/NotoSans-Regular.ttf"

# Load the font using TTFont
font = TTFont(font_path)

# Retrieve the font family name
family_name = None
for record in font["name"].names:
    if record.nameID == 1:
        family_name = record.string.decode(record.getEncoding())

# Retrieve the font style
style = None
for record in font["name"].names:
    if record.nameID == 2:
        style = record.string.decode(record.getEncoding())

# Print the font information
print("Family Name:", family_name)
print("Style:", style)

Family Name: Noto Sans
Style: Regular


# 2. Arial Unicode MS

Repeat the above, but with a different font.

In [5]:
# Specify the path to the font file
font_path = "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Arial Unicode MS Font.ttf"

# Load the font using TTFont
font = TTFont(font_path)

# Retrieve the supported Unicode character range
supported_range = font.getBestCmap().keys()

# # Print the supported Unicode characters
# for char in supported_range:
#     print(hex(char))

In [6]:
# Specify the path to the font file
font_path = "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Arial Unicode MS Font.ttf"

# Load the font using TTFont
font = TTFont(font_path)

# Retrieve the font family name
family_name = None
for record in font["name"].names:
    if record.nameID == 1:
        family_name = record.string.decode(record.getEncoding())

# Retrieve the font style
style = None
for record in font["name"].names:
    if record.nameID == 2:
        style = record.string.decode(record.getEncoding())

# Print the font information
print("Family Name:", family_name)
print("Style:", style)

Family Name: Arial Unicode MS
Style: Normal


# Comparison

In [7]:
# Specify the path to the NotoSans font file
noto_path = "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/NotoSans-Regular.ttf"

# Load the NotoSans font using TTFont
noto_font = TTFont(noto_path)

# Retrieve the supported Unicode character range for NotoSans
noto_supported_range = set(noto_font.getBestCmap().keys())

# Specify the path to the Arial Unicode MS font file
arial_path = "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Arial Unicode MS Font.ttf"

# Load the Arial Unicode MS font using TTFont
arial_font = TTFont(arial_path)

# Retrieve the supported Unicode character range for Arial Unicode MS
arial_supported_range = set(arial_font.getBestCmap().keys())

# Find unique Unicode characters in NotoSans
noto_unique_chars = noto_supported_range - arial_supported_range

# Find unique Unicode characters in Arial Unicode MS
arial_unique_chars = arial_supported_range - noto_supported_range

# Find overlapping Unicode characters
overlap_chars = noto_supported_range.intersection(arial_supported_range)

# Print the results:

# print("Unique characters in NotoSans:")
# for char in noto_unique_chars:
#     print(hex(char))

# print("\nUnique characters in Arial Unicode MS:")
# for char in arial_unique_chars:
#     print(hex(char))

# print("\nOverlap characters:")
# for char in overlap_chars:
#     print(hex(char))

Alternatively, we can just define a function to compare the Unicode coverage between multiple fonts.

In [8]:
def compare_unicode_coverage(font_paths):
    unicode_coverage = {}

    for font_path in font_paths:
        font = TTFont(font_path)
        supported_range = set(font.getBestCmap().keys())
        unicode_coverage[font_path] = supported_range

    overlap_chars = set.intersection(*unicode_coverage.values())
    unique_chars = {font_path: supported_range - overlap_chars for font_path, 
                    supported_range in unicode_coverage.items()}

    return overlap_chars, unique_chars

In [9]:
# Example usage comparing NotoSans, Arial Unicode MS, Courier New, Segoe UI, and Times New Roman:
font_paths = ["/Users/fatimaadmin/Documents/Alphabets/data/ttfs/NotoSans-Regular.ttf",
              "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Arial Unicode MS Font.ttf",
              "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/courier_new.ttf",
              "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Segoe UI.ttf",
              "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Times New Roman Font.ttf",
              "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/CODE2000.TTF"]
overlap_chars, unique_chars = compare_unicode_coverage(font_paths)

# print("Overlap characters:")
# for char in overlap_chars:
#     print(hex(char))

# print("\nUnique characters:")
# for font_path, chars in unique_chars.items():
#     print(f"\n{font_path}:")
#     for char in chars:
#         print(hex(char))

The hexadecimal representation is used to represent Unicode characters in a compact and standardized way. It provides a numerical representation of the Unicode code point, allowing easy reference to specific characters.

### Unique Characters:

Under each font path, there are multiple lines representing the unique characters in hexadecimal format. Each line corresponds to a unique character.

### Overlapping Characters:
Each line represents an overlap character, i.e., a character that is present in the Unicode coverage of multiple fonts.

This information can help us identify common symbols or characters that are widely supported across different fonts. Thus we can determine the essential characters that should be covered by any font we choose.

# Filtering Fonts

In [10]:
def filter_fonts_by_coverage(font_paths, desired_characters):
    filtered_fonts = []

    for font_path in font_paths:
        font = TTFont(font_path)
        supported_range = set(font.getBestCmap().keys())

        overlap_chars = supported_range.intersection(desired_characters)
        unique_chars = supported_range - overlap_chars
        
# Check if the font has at least 1 overlap character and 10 unique characters 
# This can be adjusted according to our requirements
        if len(overlap_chars) >= 3 and len(unique_chars) >= 10:
            filtered_fonts.append(font_path)

    return filtered_fonts

# Example usage:
font_paths = [
"/Users/fatimaadmin/Documents/Alphabets/data/ttfs/NotoSans-Regular.ttf",
"/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Arial Unicode MS Font.ttf",
"/Users/fatimaadmin/Documents/Alphabets/data/ttfs/courier_new.ttf",
"/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Segoe UI.ttf",
"/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Times New Roman Font.ttf",
"/Users/fatimaadmin/Documents/Alphabets/data/ttfs/CODE2000.TTF"
]
desired_characters = {
    0x0020,  # Space
    0x0021,  # Exclamation mark
    0x0022,  # Quotation mark
    # Add more desired characters here
}

filtered_fonts = filter_fonts_by_coverage(font_paths, desired_characters)

print("Filtered Fonts:")
for font_path in filtered_fonts:
    print(font_path)

Filtered Fonts:
/Users/fatimaadmin/Documents/Alphabets/data/ttfs/NotoSans-Regular.ttf
/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Arial Unicode MS Font.ttf
/Users/fatimaadmin/Documents/Alphabets/data/ttfs/courier_new.ttf
/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Segoe UI.ttf
/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Times New Roman Font.ttf
/Users/fatimaadmin/Documents/Alphabets/data/ttfs/CODE2000.TTF


# Incorrect/Unsupported Character Representations in Fonts

Fonts are designed to provide visual representations of characters in a specific writing system. Each font has a defined set of supported characters, and it assigns visual shapes or glyphs to those characters. However, it's possible that some characters are missing or not properly represented in a font.

In [11]:
def check_unsupported_chars(font_path, desired_characters):
    font = TTFont(font_path)
    supported_range = set(font.getBestCmap().keys())
    unsupported_chars = [] # Empty list 

    for char_code in desired_characters:
        if char_code not in supported_range:
            unsupported_chars.append(chr(char_code))

    return unsupported_chars

# Example:
font_path = "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Arial Unicode MS Font.ttf"
# List of desired characters to check
desired_characters = [0x0020, 0x0021, 0x0022, 0x066F, 0x035D, 0x03F8, 0x061E, 0x06B8, 0x0020]  

unsupported_chars = check_unsupported_chars(font_path, desired_characters)
if unsupported_chars:
    print("Unsupported representations found for the following characters:")
    for char in unsupported_chars:
        print(char)
else:
    print("No unsupported characters found.")

Unsupported representations found for the following characters:
ٯ
͝
ϸ
؞
ڸ


Update version of the above so that we can specify the Unicode range (start and end code points) we want to check. The function 'check_unsupported_chars' iterates over that range to find unsupported characters in the font. This way, we don't need to manually input specific characters to check, and it covers a wider range of characters automatically.

In [19]:
def check_unsupported_chars(font_path, unicode_range):
    '''Checks for unsupported characters within a specified unicode range for a given font'''
    font = TTFont(font_path)
    supported_range = set(font.getBestCmap().keys())
    unsupported_chars = []

    for char_code in range(unicode_range[0], unicode_range[1]+1):
        if char_code not in supported_range:
            unsupported_chars.append(chr(char_code))

    return unsupported_chars

# Example for Arial Unicode MS:
font_path = "/Users/fatimaadmin/Documents/Alphabets/data/ttfs/Arial Unicode MS Font.ttf"
unicode_range = (0x0000, 0xFFFF)  # Unicode range to check, from U+0000 to U+FFFF

unsupported_chars = check_unsupported_chars(font_path, unicode_range)
if unsupported_chars:
    print("Unsupported representations found for the following characters:")
    for char in unsupported_chars:
        print(char)
else:
    print("No unsupported characters found.")

Unsupported representations found for the following characters:









	
























€

‚
ƒ
„
…
†
‡
ˆ
‰
Š
‹
Œ

Ž


‘
’
“
”
•
–
—
˜
™
š
›
œ

ž
Ÿ
Ƕ
Ƿ
Ǹ
ǹ
Ș
ș
Ț
ț
Ȝ
ȝ
Ȟ
ȟ
Ƞ
ȡ
Ȣ
ȣ
Ȥ
ȥ
Ȧ
ȧ
Ȩ
ȩ
Ȫ
ȫ
Ȭ
ȭ
Ȯ
ȯ
Ȱ
ȱ
Ȳ
ȳ
ȴ
ȵ
ȶ
ȷ
ȸ
ȹ
Ⱥ
Ȼ
ȼ
Ƚ
Ⱦ
ȿ
ɀ
Ɂ
ɂ
Ƀ
Ʉ
Ʌ
Ɇ
ɇ
Ɉ
ɉ
Ɋ
ɋ
Ɍ
ɍ
Ɏ
ɏ
ʩ
ʪ
ʫ
ʬ
ʭ
ʮ
ʯ
˟
˪
˫
ˬ
˭
ˮ
˯
˰
˱
˲
˳
˴
˵
˶
˷
˸
˹
˺
˻
˼
˽
˾
˿
͆
͇
͈
͉
͊
͋
͌
͍
͎
͏
͐
͑
͒
͓
͔
͕
͖
͗
͘
͙
͚
͛
͜
͝
͞
͟
͢
ͣ
ͤ
ͥ
ͦ
ͧ
ͨ
ͩ
ͪ
ͫ
ͬ
ͭ
ͮ
ͯ
Ͱ
ͱ
Ͳ
ͳ
Ͷ
ͷ
͸
͹
ͻ
ͼ
ͽ
Ϳ
΀
΁
΂
΃
΋
΍
΢
Ϗ
ϗ
Ϙ
ϙ
ϛ
ϝ
ϟ
ϡ
ϴ
ϵ
϶
Ϸ
ϸ
Ϲ
Ϻ
ϻ
ϼ
Ͻ
Ͼ
Ͽ
Ѐ
Ѝ
ѐ
ѝ
҇
҈
҉
Ҋ
ҋ
Ҍ
ҍ
Ҏ
ҏ
Ӆ
ӆ
Ӊ
ӊ
Ӎ
ӎ
ӏ
Ӭ
ӭ
Ӷ
ӷ
Ӻ
ӻ
Ӽ
ӽ
Ӿ
ӿ
Ԁ
ԁ
Ԃ
ԃ
Ԅ
ԅ
Ԇ
ԇ
Ԉ
ԉ
Ԋ
ԋ
Ԍ
ԍ
Ԏ
ԏ
Ԑ
ԑ
Ԓ
ԓ
Ԕ
ԕ
Ԗ
ԗ
Ԙ
ԙ
Ԛ
ԛ
Ԝ
ԝ
Ԟ
ԟ
Ԡ
ԡ
Ԣ
ԣ
Ԥ
ԥ
Ԧ
ԧ
Ԩ
ԩ
Ԫ
ԫ
Ԭ
ԭ
Ԯ
ԯ
԰
՗
՘
ՠ
ֈ
֊
֋
֌
֍
֎
֏
֐
֢
ֺ
ׅ
׆
ׇ
׈
׉
׊
׋
׌
׍
׎
׏
׫
׬
׭
׮
ׯ
׵
׶
׷
׸
׹
׺
׻
׼
׽
׾
׿
؀
؁
؂
؃
؄
؅
؆
؇
؈
؉
؊
؋
؍
؎
؏
ؐ
ؑ
ؒ
ؓ
ؔ
ؕ
ؖ
ؗ
ؘ
ؙ
ؚ
؜
؝
؞
ؠ
ػ
ؼ
ؽ
ؾ
ؿ
ٓ
ٔ
ٕ
ٖ
ٗ
٘
ٙ
ٚ
ٛ
ٜ
ٝ
ٞ
ٟ
ٮ
ٯ
ڸ
ڹ
ڿ
ۏ
ۮ
ۯ
ۺ
ۻ
ۼ
۽
۾
ۿ
܀
܁
܂
܃
܄
܅
܆
܇
܈
܉
܊
܋
܌
܍
܎
܏
ܐ
ܑ
ܒ
ܓ
ܔ
ܕ
ܖ
ܗ
ܘ
ܙ
ܚ
ܛ
ܜ
ܝ
ܞ
ܟ
ܠ
ܡ
ܢ
ܣ
ܤ
ܥ
ܦ
ܧ
ܨ
ܩ
ܪ
ܫ
ܬ
ܭ
ܮ
ܯ
ܰ
ܱ
ܲ
ܳ
ܴ
ܵ

᫓
᫔
᫕
᫖
᫗
᫘
᫙
᫚
᫛
᫜
᫝
᫞
᫟
᫠
᫡
᫢
᫣
᫤
᫥
᫦
᫧
᫨
᫩
᫪
᫫
᫬
᫭
᫮
᫯
᫰
᫱
᫲
᫳
᫴
᫵
᫶
᫷
᫸
᫹
᫺
᫻
᫼
᫽
᫾
᫿
ᬀ
ᬁ
ᬂ
ᬃ
ᬄ
ᬅ
ᬆ
ᬇ
ᬈ
ᬉ
ᬊ
ᬋ
ᬌ
ᬍ
ᬎ
ᬏ
ᬐ
ᬑ
ᬒ
ᬓ
ᬔ
ᬕ
ᬖ
ᬗ
ᬘ
ᬙ
ᬚ
ᬛ
ᬜ
ᬝ
ᬞ
ᬟ
ᬠ
ᬡ
ᬢ
ᬣ
ᬤ
ᬥ
ᬦ
ᬧ
ᬨ
ᬩ
ᬪ
ᬫ
ᬬ
ᬭ
ᬮ
ᬯ
ᬰ
ᬱ
ᬲ
ᬳ
᬴
ᬵ
ᬶ
ᬷ
ᬸ
ᬹ
ᬺ
ᬻ
ᬼ
ᬽ
ᬾ
ᬿ
ᭀ
ᭁ
ᭂ
ᭃ
᭄
ᭅ
ᭆ
ᭇ
ᭈ
ᭉ
ᭊ
ᭋ
ᭌ
᭍
᭎
᭏
᭐
᭑
᭒
᭓
᭔
᭕
᭖
᭗
᭘
᭙
᭚
᭛
᭜
᭝
᭞
᭟
᭠
᭡
᭢
᭣
᭤
᭥
᭦
᭧
᭨
᭩
᭪
᭫
᭬
᭭
᭮
᭯
᭰
᭱
᭲
᭳
᭴
᭵
᭶
᭷
᭸
᭹
᭺
᭻
᭼
᭽
᭾
᭿
ᮀ
ᮁ
ᮂ
ᮃ
ᮄ
ᮅ
ᮆ
ᮇ
ᮈ
ᮉ
ᮊ
ᮋ
ᮌ
ᮍ
ᮎ
ᮏ
ᮐ
ᮑ
ᮒ
ᮓ
ᮔ
ᮕ
ᮖ
ᮗ
ᮘ
ᮙ
ᮚ
ᮛ
ᮜ
ᮝ
ᮞ
ᮟ
ᮠ
ᮡ
ᮢ
ᮣ
ᮤ
ᮥ
ᮦ
ᮧ
ᮨ
ᮩ
᮪
᮫
ᮬ
ᮭ
ᮮ
ᮯ
᮰
᮱
᮲
᮳
᮴
᮵
᮶
᮷
᮸
᮹
ᮺ
ᮻ
ᮼ
ᮽ
ᮾ
ᮿ
ᯀ
ᯁ
ᯂ
ᯃ
ᯄ
ᯅ
ᯆ
ᯇ
ᯈ
ᯉ
ᯊ
ᯋ
ᯌ
ᯍ
ᯎ
ᯏ
ᯐ
ᯑ
ᯒ
ᯓ
ᯔ
ᯕ
ᯖ
ᯗ
ᯘ
ᯙ
ᯚ
ᯛ
ᯜ
ᯝ
ᯞ
ᯟ
ᯠ
ᯡ
ᯢ
ᯣ
ᯤ
ᯥ
᯦
ᯧ
ᯨ
ᯩ
ᯪ
ᯫ
ᯬ
ᯭ
ᯮ
ᯯ
ᯰ
ᯱ
᯲
᯳
᯴
᯵
᯶
᯷
᯸
᯹
᯺
᯻
᯼
᯽
᯾
᯿
ᰀ
ᰁ
ᰂ
ᰃ
ᰄ
ᰅ
ᰆ
ᰇ
ᰈ
ᰉ
ᰊ
ᰋ
ᰌ
ᰍ
ᰎ
ᰏ
ᰐ
ᰑ
ᰒ
ᰓ
ᰔ
ᰕ
ᰖ
ᰗ
ᰘ
ᰙ
ᰚ
ᰛ
ᰜ
ᰝ
ᰞ
ᰟ
ᰠ
ᰡ
ᰢ
ᰣ
ᰤ
ᰥ
ᰦ
ᰧ
ᰨ
ᰩ
ᰪ
ᰫ
ᰬ
ᰭ
ᰮ
ᰯ
ᰰ
ᰱ
ᰲ
ᰳ
ᰴ
ᰵ
ᰶ
᰷
᰸
᰹
᰺
᰻
᰼
᰽
᰾
᰿
᱀
᱁
᱂
᱃
᱄
᱅
᱆
᱇
᱈
᱉
᱊
᱋
᱌
ᱍ
ᱎ
ᱏ
᱐
᱑
᱒
᱓
᱔
᱕
᱖
᱗
᱘
᱙
ᱚ
ᱛ
ᱜ
ᱝ
ᱞ
ᱟ
ᱠ
ᱡ
ᱢ
ᱣ
ᱤ
ᱥ
ᱦ
ᱧ
ᱨ
ᱩ
ᱪ
ᱫ
ᱬ
ᱭ
ᱮ
ᱯ
ᱰ
ᱱ
ᱲ
ᱳ
ᱴ
ᱵ
ᱶ
ᱷ
ᱸ
ᱹ
ᱺ
ᱻ
ᱼ
ᱽ
᱾
᱿
ᲀ
ᲁ
ᲂ
ᲃ
ᲄ
ᲅ
ᲆ
ᲇ
ᲈ
Ᲊ
ᲊ
᲋
᲌
᲍
᲎
᲏
Ა
Ბ
Გ
Დ
Ე
Ვ
Ზ
Თ
Ი
Კ
Ლ
Მ
Ნ
Ო
Პ
Ჟ
Რ
Ს
Ტ
Უ
Ფ
Ქ
Ღ
Ყ
Შ
Ჩ
Ც
Ძ
Წ
Ჭ
Ხ
Ჯ
Ჰ
Ჱ
Ჲ
Ჳ
Ჴ
Ჵ
Ჶ
Ჷ
Ჸ
Ჹ
Ჺ
᲻
᲼
Ჽ
Ჾ
Ჿ
᳀
᳁
᳂
᳃
᳄
᳅
᳆


㰺
㰻
㰼
㰽
㰾
㰿
㱀
㱁
㱂
㱃
㱄
㱅
㱆
㱇
㱈
㱉
㱊
㱋
㱌
㱍
㱎
㱏
㱐
㱑
㱒
㱓
㱔
㱕
㱖
㱗
㱘
㱙
㱚
㱛
㱜
㱝
㱞
㱟
㱠
㱡
㱢
㱣
㱤
㱥
㱦
㱧
㱨
㱩
㱪
㱫
㱬
㱭
㱮
㱯
㱰
㱱
㱲
㱳
㱴
㱵
㱶
㱷
㱸
㱹
㱺
㱻
㱼
㱽
㱾
㱿
㲀
㲁
㲂
㲃
㲄
㲅
㲆
㲇
㲈
㲉
㲊
㲋
㲌
㲍
㲎
㲏
㲐
㲑
㲒
㲓
㲔
㲕
㲖
㲗
㲘
㲙
㲚
㲛
㲜
㲝
㲞
㲟
㲠
㲡
㲢
㲣
㲤
㲥
㲦
㲧
㲨
㲩
㲪
㲫
㲬
㲭
㲮
㲯
㲰
㲱
㲲
㲳
㲴
㲵
㲶
㲷
㲸
㲹
㲺
㲻
㲼
㲽
㲾
㲿
㳀
㳁
㳂
㳃
㳄
㳅
㳆
㳇
㳈
㳉
㳊
㳋
㳌
㳍
㳎
㳏
㳐
㳑
㳒
㳓
㳔
㳕
㳖
㳗
㳘
㳙
㳚
㳛
㳜
㳝
㳞
㳟
㳠
㳡
㳢
㳣
㳤
㳥
㳦
㳧
㳨
㳩
㳪
㳫
㳬
㳭
㳮
㳯
㳰
㳱
㳲
㳳
㳴
㳵
㳶
㳷
㳸
㳹
㳺
㳻
㳼
㳽
㳾
㳿
㴀
㴁
㴂
㴃
㴄
㴅
㴆
㴇
㴈
㴉
㴊
㴋
㴌
㴍
㴎
㴏
㴐
㴑
㴒
㴓
㴔
㴕
㴖
㴗
㴘
㴙
㴚
㴛
㴜
㴝
㴞
㴟
㴠
㴡
㴢
㴣
㴤
㴥
㴦
㴧
㴨
㴩
㴪
㴫
㴬
㴭
㴮
㴯
㴰
㴱
㴲
㴳
㴴
㴵
㴶
㴷
㴸
㴹
㴺
㴻
㴼
㴽
㴾
㴿
㵀
㵁
㵂
㵃
㵄
㵅
㵆
㵇
㵈
㵉
㵊
㵋
㵌
㵍
㵎
㵏
㵐
㵑
㵒
㵓
㵔
㵕
㵖
㵗
㵘
㵙
㵚
㵛
㵜
㵝
㵞
㵟
㵠
㵡
㵢
㵣
㵤
㵥
㵦
㵧
㵨
㵩
㵪
㵫
㵬
㵭
㵮
㵯
㵰
㵱
㵲
㵳
㵴
㵵
㵶
㵷
㵸
㵹
㵺
㵻
㵼
㵽
㵾
㵿
㶀
㶁
㶂
㶃
㶄
㶅
㶆
㶇
㶈
㶉
㶊
㶋
㶌
㶍
㶎
㶏
㶐
㶑
㶒
㶓
㶔
㶕
㶖
㶗
㶘
㶙
㶚
㶛
㶜
㶝
㶞
㶟
㶠
㶡
㶢
㶣
㶤
㶥
㶦
㶧
㶨
㶩
㶪
㶫
㶬
㶭
㶮
㶯
㶰
㶱
㶲
㶳
㶴
㶵
㶶
㶷
㶸
㶹
㶺
㶻
㶼
㶽
㶾
㶿
㷀
㷁
㷂
㷃
㷄
㷅
㷆
㷇
㷈
㷉
㷊
㷋
㷌
㷍
㷎
㷏
㷐
㷑
㷒
㷓
㷔
㷕
㷖
㷗
㷘
㷙
㷚
㷛
㷜
㷝
㷞
㷟
㷠
㷡
㷢
㷣
㷤
㷥
㷦
㷧
㷨
㷩
㷪
㷫
㷬
㷭
㷮
㷯
㷰
㷱
㷲
㷳
㷴
㷵
㷶
㷷
㷸
㷹
㷺
㷻
㷼
㷽
㷾
㷿
㸀
㸁
㸂
㸃
㸄
㸅
㸆
㸇
㸈
㸉
㸊
㸋
㸌
㸍
㸎
㸏
㸐
㸑
㸒
㸓
㸔
㸕
㸖
㸗
㸘
㸙
㸚
㸛
㸜
㸝
㸞
㸟
㸠
㸡
㸢
㸣
㸤
㸥
㸦
㸧
㸨
㸩
㸪
㸫
㸬
㸭


䷍
䷎
䷏
䷐
䷑
䷒
䷓
䷔
䷕
䷖
䷗
䷘
䷙
䷚
䷛
䷜
䷝
䷞
䷟
䷠
䷡
䷢
䷣
䷤
䷥
䷦
䷧
䷨
䷩
䷪
䷫
䷬
䷭
䷮
䷯
䷰
䷱
䷲
䷳
䷴
䷵
䷶
䷷
䷸
䷹
䷺
䷻
䷼
䷽
䷾
䷿
龦
龧
龨
龩
龪
龫
龬
龭
龮
龯
龰
龱
龲
龳
龴
龵
龶
龷
龸
龹
龺
龻
龼
龽
龾
龿
鿀
鿁
鿂
鿃
鿄
鿅
鿆
鿇
鿈
鿉
鿊
鿋
鿌
鿍
鿎
鿏
鿐
鿑
鿒
鿓
鿔
鿕
鿖
鿗
鿘
鿙
鿚
鿛
鿜
鿝
鿞
鿟
鿠
鿡
鿢
鿣
鿤
鿥
鿦
鿧
鿨
鿩
鿪
鿫
鿬
鿭
鿮
鿯
鿰
鿱
鿲
鿳
鿴
鿵
鿶
鿷
鿸
鿹
鿺
鿻
鿼
鿽
鿾
鿿
ꀀ
ꀁ
ꀂ
ꀃ
ꀄ
ꀅ
ꀆ
ꀇ
ꀈ
ꀉ
ꀊ
ꀋ
ꀌ
ꀍ
ꀎ
ꀏ
ꀐ
ꀑ
ꀒ
ꀓ
ꀔ
ꀕ
ꀖ
ꀗ
ꀘ
ꀙ
ꀚ
ꀛ
ꀜ
ꀝ
ꀞ
ꀟ
ꀠ
ꀡ
ꀢ
ꀣ
ꀤ
ꀥ
ꀦ
ꀧ
ꀨ
ꀩ
ꀪ
ꀫ
ꀬ
ꀭ
ꀮ
ꀯ
ꀰ
ꀱ
ꀲ
ꀳ
ꀴ
ꀵ
ꀶ
ꀷ
ꀸ
ꀹ
ꀺ
ꀻ
ꀼ
ꀽ
ꀾ
ꀿ
ꁀ
ꁁ
ꁂ
ꁃ
ꁄ
ꁅ
ꁆ
ꁇ
ꁈ
ꁉ
ꁊ
ꁋ
ꁌ
ꁍ
ꁎ
ꁏ
ꁐ
ꁑ
ꁒ
ꁓ
ꁔ
ꁕ
ꁖ
ꁗ
ꁘ
ꁙ
ꁚ
ꁛ
ꁜ
ꁝ
ꁞ
ꁟ
ꁠ
ꁡ
ꁢ
ꁣ
ꁤ
ꁥ
ꁦ
ꁧ
ꁨ
ꁩ
ꁪ
ꁫ
ꁬ
ꁭ
ꁮ
ꁯ
ꁰ
ꁱ
ꁲ
ꁳ
ꁴ
ꁵ
ꁶ
ꁷ
ꁸ
ꁹ
ꁺ
ꁻ
ꁼ
ꁽ
ꁾ
ꁿ
ꂀ
ꂁ
ꂂ
ꂃ
ꂄ
ꂅ
ꂆ
ꂇ
ꂈ
ꂉ
ꂊ
ꂋ
ꂌ
ꂍ
ꂎ
ꂏ
ꂐ
ꂑ
ꂒ
ꂓ
ꂔ
ꂕ
ꂖ
ꂗ
ꂘ
ꂙ
ꂚ
ꂛ
ꂜ
ꂝ
ꂞ
ꂟ
ꂠ
ꂡ
ꂢ
ꂣ
ꂤ
ꂥ
ꂦ
ꂧ
ꂨ
ꂩ
ꂪ
ꂫ
ꂬ
ꂭ
ꂮ
ꂯ
ꂰ
ꂱ
ꂲ
ꂳ
ꂴ
ꂵ
ꂶ
ꂷ
ꂸ
ꂹ
ꂺ
ꂻ
ꂼ
ꂽ
ꂾ
ꂿ
ꃀ
ꃁ
ꃂ
ꃃ
ꃄ
ꃅ
ꃆ
ꃇ
ꃈ
ꃉ
ꃊ
ꃋ
ꃌ
ꃍ
ꃎ
ꃏ
ꃐ
ꃑ
ꃒ
ꃓ
ꃔ
ꃕ
ꃖ
ꃗ
ꃘ
ꃙ
ꃚ
ꃛ
ꃜ
ꃝ
ꃞ
ꃟ
ꃠ
ꃡ
ꃢ
ꃣ
ꃤ
ꃥ
ꃦ
ꃧ
ꃨ
ꃩ
ꃪ
ꃫ
ꃬ
ꃭ
ꃮ
ꃯ
ꃰ
ꃱ
ꃲ
ꃳ
ꃴ
ꃵ
ꃶ
ꃷ
ꃸ
ꃹ
ꃺ
ꃻ
ꃼ
ꃽ
ꃾ
ꃿ
ꄀ
ꄁ
ꄂ
ꄃ
ꄄ
ꄅ
ꄆ
ꄇ
ꄈ
ꄉ
ꄊ
ꄋ
ꄌ
ꄍ
ꄎ
ꄏ
ꄐ
ꄑ
ꄒ
ꄓ
ꄔ
ꄕ
ꄖ
ꄗ
ꄘ
ꄙ
ꄚ
ꄛ
ꄜ
ꄝ
ꄞ
ꄟ
ꄠ
ꄡ
ꄢ
ꄣ
ꄤ
ꄥ
ꄦ
ꄧ
ꄨ
ꄩ
ꄪ
ꄫ
ꄬ
ꄭ
ꄮ
ꄯ
ꄰ
ꄱ
ꄲ
ꄳ
ꄴ
ꄵ
ꄶ
ꄷ
ꄸ
ꄹ
ꄺ
ꄻ
ꄼ
ꄽ
ꄾ
ꄿ
ꅀ
ꅁ
ꅂ
ꅃ
ꅄ
ꅅ
ꅆ
ꅇ
ꅈ
ꅉ
ꅊ
ꅋ
ꅌ
ꅍ
ꅎ
ꅏ
ꅐ
ꅑ
ꅒ
ꅓ
ꅔ
ꅕ
ꅖ
ꅗ
ꅘ
ꅙ
ꅚ
ꅛ
ꅜ
ꅝ
ꅞ
ꅟ
ꅠ
ꅡ
ꅢ
ꅣ
ꅤ
ꅥ
ꅦ
























































































































































































































































































































































































































































































































In [16]:
def count_unsupported_chars(font_path, unicode_range):
    '''Counts the number of unsupported characters within a specified unicode range for a given font'''
    font = TTFont(font_path)
    supported_range = set(font.getBestCmap().keys())
    unsupported_chars = sum(char_code not in supported_range for char_code in range(unicode_range[0], unicode_range[1] + 1))
    return unsupported_chars

folder_path = "/Users/fatimaadmin/Documents/Alphabets/data/ttfs"  # Path to the folder containing font files
unicode_range = (0x0000, 0xFFFF)  # Unicode range to check, from U+0000 to U+FFFF

font_files = []
for root, dirs, files in os.walk(folder_path):
    for file in files:
        if file.endswith((".ttf", ".otf")):
            font_files.append(os.path.join(root, file))

for font_file in font_files:
    font_name = os.path.splitext(os.path.basename(font_file))[0]
    unsupported_chars = count_unsupported_chars(font_file, unicode_range)
    print(f"Font: {font_name}")
    print(f"Unsupported characters: {unsupported_chars}")
    print()


Font: NotoSans-Regular
Unsupported characters: 62483

Font: Arial Unicode MS Font
Unsupported characters: 26619

Font: Times New Roman Font
Unsupported characters: 65320

Font: Segoe UI
Unsupported characters: 65344

Font: courier_new
Unsupported characters: 62690

Font: GNU Unifont
Unsupported characters: 8450

Font: CODE2000
Unsupported characters: 12750

Font: DejaVuSansCondensed-BoldOblique
Unsupported characters: 60717

Font: DejaVuMathTeXGyre
Unsupported characters: 64418

Font: DejaVuSansCondensed
Unsupported characters: 60166

Font: DejaVuSansCondensed-Bold
Unsupported characters: 60170

Font: DejaVuSerifCondensed-Bold
Unsupported characters: 62215

Font: DejaVuSerif-BoldItalic
Unsupported characters: 62211

Font: DejaVuSerifCondensed-BoldItalic
Unsupported characters: 62211

Font: DejaVuSans-BoldOblique
Unsupported characters: 60717

Font: DejaVuSans
Unsupported characters: 60166

Font: DejaVuSansCondensed-Oblique
Unsupported characters: 60717

Font: DejaVuSerifCondensed-Itali

Fonts with the least unsupported characters so far are GNU Unifont (8450), followed by Code2000 (12750).